### PR TITLE
Enterprise GitHub using Personal Tokens

### PR DESCRIPTION
**Pull request by rcaileff**
GitHub Enterprise Support #99 version 2.

I haven't updated my team's server to use this version because I don't want to disrupt our existing CI.  I did test this out on my laptop as far as the BuildTemplate selection step.  I had to use a hardcoded change to a string value in GitServerFactory.swift because of a quirk in our environment; I don't know if other users will have a similar problem.

### CODE REVIEW

**Comment by rcaileff** [comment]
The "Fix this" commit has been backed out as part of the "Implement" commit.  I don't know why Github insists there is a conflict.

File: BuildaGitServer/GitServerPublic.swift [hunk 1]
```diff
@@ -41,6 +53,7 @@ public enum GitService: String {
     public func authorizeUrl() -> String {
         switch self {
         case .GitHub: return "https://github.com/login/oauth/authorize"
+        case .EnterpriseGitHub: return "https://\(hostname())/login/oauth/authorize"
```
Comment (czechboy0) [inline BuildaGitServer/GitServerPublic.swift]: I think we should have an assert here as well, since we don't support GitHub Enterprise OAuth yet.

**Comment by czechboy0** [comment]
Can you please merge master into your branch again? At the moment it can't be merged.

As I didn't manage to get a local GitHub Enterprise to run locally (even though I have a license), I'll rely on you testing it before we merge it. I won't have a chance to verify it.

**Comment by rcaileff** [comment]
I still get build errors from master, but here's a merged version for you.

**Comment by rcaileff** [comment]
Merged.

File: Buildasaur/Base.lproj/Main.storyboard [hunk 2]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.Cocoa.Storyboard.XIB" version="3.0" toolsVersion="10089" systemVersion="15E33e" targetRuntime="MacOSX.Cocoa" propertyAccessControl="none" useAutolayout="YES">
+<document type="com.apple.InterfaceBuilder3.Cocoa.Storyboard.XIB" version="3.0" toolsVersion="10109" systemVersion="15D21" targetRuntime="MacOSX.Cocoa" propertyAccessControl="none" useAutolayout="YES">
     <dependencies>
         <deployment identifier="macosx"/>
-        <plugIn identifier="com.apple.InterfaceBuilder.CocoaPlugin" version="10089"/>
+        <plugIn identifier="com.apple.InterfaceBuilder.CocoaPlugin" version="10109"/>
```
Comment (czechboy0) [inline Buildasaur/Base.lproj/Main.storyboard]: Can you please revert these changes, since you didn't change anything I'd prefer to not have a diff here.

File: BuildaGitServerTests/EnterpriseGitHubSourceTests.swift [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        let request = try! self.github.endpoints.createRequest(method, endpoint: endpoint, params: params)
+
+        self.github.http.sendRequest(request, completion: { (response, body, error) -> () in
+
+            completion(body: body, error: error)
+            expect.fulfill()
+        })
+
+        waitForExpectationsWithTimeout(10, handler: nil)
```
Comment (czechboy0) [inline BuildaGitServerTests/EnterpriseGitHubSourceTests.swift]: Isn't this going to fail outside of your network?

File: BuildaGitServer/Base/Authentication.swift [hunk 4]
```diff
@@ -40,9 +40,17 @@ extension ProjectAuthenticator: KeychainStringSerializable {
...
-            throw Error.withInfo("Unsupported service: \(comps[0])")
+
+        var service: GitService
+        switch comps[0] {
+        case GitService.GitHub.hostname():
+            service = GitService.GitHub
+        case GitService.BitBucket.hostname():
+            service = GitService.BitBucket
+        default:
+            service = GitService.EnterpriseGitHub(host: comps[0])
```
Comment (czechboy0) [inline BuildaGitServer/Base/Authentication.swift]: Hmm this will be tricky when someone tries to add a service not on any supported server. Could you instead try to synchronously ping the server and detect whether it really is a GitHub Enterprise server? We'll need this anyway to support more self-hosted servers like BitBucket Server - we'll need to ping the parsed host and detect which type it is. So using the `default` branch of the switch is here might cause issues with error reporting (we'll let users add even completely unsupported servers, which I'd like to avoid).

File: BuildaGitServerTests/EnterpriseGitHubSourceTests.swift (hunk 3 above)
Comment (rcaileff) [inline BuildaGitServerTests/EnterpriseGitHubSourceTests.swift]: The tests are commented out.  The helper functions shouldn't be a problem.

File: Buildasaur/Base.lproj/Main.storyboard (hunk 2 above)
Comment (rcaileff) [inline Buildasaur/Base.lproj/Main.storyboard]: Done.

File: BuildaGitServer/Base/Authentication.swift (hunk 4 above)
Comment (rcaileff) [inline BuildaGitServer/Base/Authentication.swift]: I've got a check for this now.  I don't know if it's robust enough for you, but it was the best I could come up with.

**Comment by fotiDim** [comment]
I tried to use @rcaileff 's branch on our GH Enterprise instance and it failed due to the ldap authentication we use. When you point the tool to the GH API URL it always gets redirected to the LDAP authentication HTML page through which you have to go before getting into GH. 

This issue would have been avoided if SSH keys (Deploy keys) authentication was used. Since this is a server-server authentication it makes much sense not to use Personal Tokens but Deploy keys instead.

File: BuildaGitServer/Base/Authentication.swift [hunk 5]
```diff
@@ -69,4 +73,16 @@ extension ProjectAuthenticator: KeychainStringSerializable {
             self.secret
             ].joinWithSeparator(":")
     }
+
+    internal static func createEnterpriseService(host: String) -> GitService? {
+        guard let url = NSURL.init(string: "http://\(host)") else { return nil }
```
Comment (czechboy0) [inline BuildaGitServer/Base/Authentication.swift]: You can just do `NSURL(string: ...)`.

File: BuildaGitServer/Base/Authentication.swift [hunk 6]
```diff
@@ -69,4 +73,16 @@ extension ProjectAuthenticator: KeychainStringSerializable {
...
+        do {
+            let response = try NSString.init(contentsOfURL: url, encoding: NSASCIIStringEncoding)
+            if response.lowercaseString.containsString("github") {
```
Comment (czechboy0) [inline BuildaGitServer/Base/Authentication.swift]: Can you walk me through what the response from Enterprise GitHub looks like? Does it return a string that says `github`?

**Comment by czechboy0** [comment]
@fotiDim Deploy keys are not yet supported, if you'd like to see support for them in Buildasaur, please create a separate ticket. I agree it would be nice to support them :) We just need to approach these changes in as small steps as possible to ensure everything keeps working.

Regarding this PR, @rcaileff do you understand the issue @fotiDim is having? Unfortunately I don't have a Enterprise server I would have access to, so I can't chime in here.

On that note, would either of you be able to create a test repo on your Enterprise server and give me access (if it's accessible from the outside network)? It would greatly help me to get involved with these Enterprise features, being able to test them and develop more features for them.

**Comment by rcaileff** [comment]
I partially understand the problem @fotiDim is having.  Buildasaur comments
in a PR to report the test results.  When it does so, it comments as the
user whose credentials are tied to the Personal Token.  When I initially
set up Buildasaur, the credentials were tied to my personal work account,
so Buildasaur was making comments as me.  This was inconvenient for several
reasons.  I'm not sure what his proposed solution involves.

@czechboy0 Unfortunately, I can't give you access to the corporate servers.

**Comment by czechboy0** [comment]
What I did to solve the first (comments coming in under my personal user) is that I created a separate GitHub account (@buildasaur) whose Personal Tokens I use (after adding it as a collaborator), so that the comments are clearly coming from a CI bot. I recommend you do the same, creating a new GitHub account takes a couple of minutes and the workflow is much nicer then.

I understand it's tricky, I was just asking whether there isn't a way to create one test repository to which I'd have access, I didn't want access to the complete server of course. But if it's impossible there's nothing we can do.

**Comment by rcaileff** [comment]
I did eventually get a separate account to tie to the comments, but because
it's an Enterprise GitHub setup I had to go through corporate channels to
get it created.

File: BuildaGitServer/Base/Authentication.swift (hunk 6 above)
Comment (rcaileff) [inline BuildaGitServer/Base/Authentication.swift]: I suspect it's customized per Enterprise GitHub installation.  For our company it's a home page that looks much like I remember www.github.com looking last year.  The html happens to include the word "github" in several places.  It's an extremely fragile way of testing for an Enterprise GitHub server, but it was the best I could come up with.

Comment (czechboy0) [inline BuildaGitServer/Base/Authentication.swift]: You're right, that's my worry. I wonder if there isn't an API call appropriate for this, maybe even trying to list all repositories on that server with `/repos` or something like that. Again, I'm not of much use because I don't have an Enterprise server at my disposal :/

File: BuildaGitServer/Base/Authentication.swift (hunk 5 above)
Comment (rcaileff) [inline BuildaGitServer/Base/Authentication.swift]: Done.

File: BuildaGitServer/Base/Authentication.swift (hunk 6 above)
Comment (rcaileff) [inline BuildaGitServer/Base/Authentication.swift]: I could do this instead, but I don't think it's better because if someone with more permissions tries to use buildasaur it'll fail. :

```
internal static func createEnterpriseService(host: String) -> GitService? {
    guard let url = NSURL(string: "http://\(host)/setup/api/settings")
        else { return nil }
    do {
        let response = try NSString(contentsOfURL: url, encoding: NSASCIIStringEncoding)
        if response.lowercaseString.containsString("not authorized") {
            return GitService.EnterpriseGitHub(host: host)
        }
    } catch {
    }
    return nil
}
```

**Comment by fotiDim** [comment]
@czechboy0 unfortunately I cannot give you access to our GitHub enterprise instance as it is accessible only in our intranet. I will be happy to test things for though so just ping me when you need sth.

I am still working on the SSH authentication issue and I hope I will have fixed it soon.

**Comment by buildasaur** [comment]
Result of [Integration 1](https://stlt.herokuapp.com/v1/xcs_deeplink/honzadvysmbpr14.home/cc6638942eeca76b8e63719bed017b1b/cc6638942eeca76b8e63719bed019ba1)
---
*Duration*: 31 seconds
*Result*: **-1 error, failing state: trigger-error**

**Comment by rcaileff** [comment]
@ Navideck-fotis How is the SSH authentication issue going?

**Comment by fotiDim** [comment]
@rcaileff Hi there. Our internal IT department has failed me. They provided no solution. We settled using a more simple workflow without the benefits of Buildasaur. However our GitHub Server is accessible from the internet now so I might be able to setup an account and a repo for you.

**Comment by czechboy0** [comment]
@Navideck-fotis 🎉

**Comment by czechboy0** [comment]
@rcaileff It looks like other PRs have made master conflict with your branch. Could you merge master into your branch, resolve conflicts and get this branch up to date? I'll be happy to merge it once we know it's working! 😊

**Comment by rcaileff** [comment]
@czechboy0 I've merged in master and made some minor updates.  I have not removed the rate-limitation requirement from the server headers or otherwise tried to handle misconfigured servers.

**Comment by czechboy0** [comment]
Cool, does this work for you locally? Can anyone else please try to run this with a GitHub Enterprise server for a day and report that everything went well please? Unfortunately I can't test it myself :(

But thanks @rcaileff, this is great stuff, once we're sure it all works I'll merge it in 🎉

**Comment by rcaileff** [comment]
@czechboy0 There are a couple of changes I have to make locally because we also have a misconfigured server... but other than that it works for me.  If there's someone out there who can test it on a different system, that would be awesome.

**Comment by buildasaur** [comment]
Result of [Integration 1](https://stlt.herokuapp.com/v1/xcs_deeplink/honzadvysmbpr14.home/8d6054e0e094992ea4fa6aa54f079476/8d6054e0e094992ea4fa6aa54f0799c4)
---
*Duration*: 1 minute and 27 seconds
*Result*: **-1 error, failing state: trigger-error**

**Comment by buildasaur** [comment]
Result of [Integration 2](https://stlt.herokuapp.com/v1/xcs_deeplink/honzadvysmbpr14.home/8d6054e0e094992ea4fa6aa54f079476/8d6054e0e094992ea4fa6aa54f07d657)
---
*Duration*: 31 seconds
*Result*: **-1 error, failing state: trigger-error**

**Comment by rcaileff** [comment]
Merged master back into the branch to resolve conflicts so that the PR is once again able to be merged.  Have you had any luck finding someone who can test this?  It'd be nice to get it merged into master so I can stop fixing conflicts.  😄

**Comment by czechboy0** [comment]
Sorry about that, unfortunately it's a large enough change that I don't feel comfortable merging it unless someone is able to give it a good test. If I had access to an enterprise server I could do it myself, however that's not the case right now :(

**Comment by fotiDim** [comment]
I will give it another go internally. If there is any update I will let you know.

**Comment by sambae** [comment]
I am also interested in this feature and will do some testing

**Comment by sambae** [comment]
@fotiDim were you able to get past the ldap authentication page issue?

**Comment by sambae** [comment]
So I was able to get up and running with a couple small changes to the code.

**Comment by fotiDim** [comment]
@sambae good to hear that you got it working. We don't use ldap but active directory and with SAML authentication in the front. I am still waiting for feedback internally.
